code/modules/reagents/reagent_containers/spray.dm

```diff
@@ -14,78 +17,59 @@
-		if(!A.reagents.total_volume && A.reagents)
-			user << "<span class='notice'>\The [A] is empty.</span>"
-			return
+/obj/item/weapon/reagent_containers/spray/afterattack(atom/A as mob|obj, mob/user as mob, var/adjacency_flag, var/click_params)
+	if (adjacency_flag && is_type_in_list(A, ignore_spray_types))
+		return
 
-		if(reagents.total_volume >= reagents.maximum_volume)
-			user << "<span class='notice'>\The [src] is full.</span>"
-			return
+	if (delay_spraying)
+		user.delayNextAttack(8)
 
-		var/trans = A.reagents.trans_to(src, A:amount_per_transfer_from_this)
-		user << "<span class='notice'>You fill \the [src] with [trans] units of the contents of \the [A].</span>"
+	if (istype(A, /obj/structure/reagent_dispensers) && adjacency_flag)
+		transfer(A, user, can_send = FALSE, can_receive = TRUE)
 		return
 
-	if(reagents.total_volume < amount_per_transfer_from_this)
+	if (!can_transfer_an_APTFT())
 		user << "<span class='notice'>\The [src] is empty!</span>"
 		return
 
-	var/obj/effect/decal/D = new/obj/effect/decal(get_turf(src))
-	D.create_reagents(amount_per_transfer_from_this)
-	reagents.trans_to(D, amount_per_transfer_from_this, 1/3)
+	// Log reagents
+	var/list/log_reagent_list = list()
 
-	D.name = "chemicals"
-	D.icon = 'icons/obj/chempuff.dmi'
+	for (var/reagent_id in logged_sprayed_reagents)
+		if (reagents.has_reagent(reagent_id))
+			log_reagent_list += "'[reagent_id]'"
 
-	D.icon += mix_color_from_reagents(D.reagents.reagent_list)
+	if (log_reagent_list.len > 0)
+		add_gamelogs(user, "sprayed {[english_list(log_reagent_list, and_text = ", ")]} with \the [src]", admin = TRUE, tp_link = TRUE)
 
-	var/turf/A_turf = get_turf(A)
+	// Override for your custom puff behaviour
+	make_puff(A, user)
 
-	spawn(0)
-		for(var/i=0, i<3, i++)
-			step_towards(D,A)
-			D.reagents.reaction(get_turf(D))
-			for(var/atom/T in get_turf(D))
-				D.reagents.reaction(T)
-
-				// When spraying against the wall, also react with the wall, but
-				// not its contents.
-				if(get_dist(D, A_turf) == 1 && A_turf.density)
-					D.reagents.reaction(A_turf)
-				sleep(2)
-			sleep(3)
-		D.reagents.Destroy()
-		D.loc = null
+/obj/item/weapon/reagent_containers/spray/attack_self(var/mob/user)
+	amount_per_transfer_from_this = (amount_per_transfer_from_this == 10 ? 5 : 10)
+	user << "<span class='notice'>You switched [amount_per_transfer_from_this == 10 ? "on" : "off"] the pressure nozzle. You'll now use [amount_per_transfer_from_this] units per spray.</span>"
 
-	playsound(get_turf(src), 'sound/effects/spray2.ogg', 50, 1, -6)
+/obj/item/weapon/reagent_containers/spray/proc/make_puff(var/atom/target, var/mob/user)
+	// Create the chemical puff
+	var/mix_color = mix_color_from_reagents(reagents.reagent_list)
+	var/obj/effect/decal/chemical_puff/D = getFromPool(/obj/effect/decal/chemical_puff, get_turf(src), mix_color, amount_per_transfer_from_this)
+	reagents.trans_to(D, amount_per_transfer_from_this, 1/3)
 
-	if(reagents.has_reagent("sacid"))

```

Comment:
Because it's preserved on line 47 better and it checks that it's spraying something properly now.